server/path_watcher.go

```diff
@@ -0,0 +1,309 @@
+}
+
+func (watcher *PathWatcher) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	select {
+	case <-time.After(time.Duration(watcher.priority*masterTTL) * time.Second):
+	case <-ctx.Done():
+		return
+	}
+
+	for {
+		err := watcher.run(ctx)
+
+		switch err {
+		case errLockFailed:
+			watcher.updateCounter(1, "lock.failed")
+			log.Debug("%s failed to acquire lock, retrying...", watcher)
+		case context.Canceled:
+			// Do nothing, normal shutdown
+		default:
+			watcher.updateCounter(1, "error")
+			log.Error("%s aborted, retrying...: %s", watcher, err)
+		}
+
+		select {
+		case <-time.After(time.Duration(watcher.priority*masterTTL+1) * time.Second):
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// run handles events on a path with a handler.
+// Returns any error or context cancel.
+// This method gets a lock on the sync backend and returns with an error when fails.
+func (watcher *PathWatcher) run(parentCtx context.Context) error {
+	watcher.updateCounter(1, "active")
+	defer watcher.updateCounter(-1, "active")
+
+	lockKey := lockPath + "/watch" + watcher.path
+	lost, err := watcher.sync.Lock(parentCtx, lockKey, false)
+	if err != nil {
+		return errLockFailed
+	}
+	defer func() {
+		// can't use the parent context, it may be already canceled
+		if err := watcher.sync.Unlock(context.Background(), lockKey); err != nil {
+			log.Warning("%s unlocking etcd failed on %s: %s", watcher, lockKey, err)
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	watcher.previousProcessedRevision = watcher.fetchStoredRevision(parentCtx)
+	eventsCh := watcher.sync.Watch(ctx, watcher.path, watcher.previousProcessedRevision+1)
+	doneCh := make(chan error, 1)
+
+	go watcher.consumeEvents(ctx, eventsCh, doneCh)
+
+	select {
+	case <-ctx.Done():
+		if err := <-doneCh; err != nil {
+			log.Error("%s consuming events failed: %s", watcher, err)
+		}
+		return ctx.Err()
+	case <-lost:
+		watcher.updateCounter(1, "lock.lost")
+		cancel()
+		if err := <-doneCh; err != nil {
+			log.Error("%s error after lost lock: %s", watcher, err)
+		}
+		return fmt.Errorf("lock for path `%s` is lost", watcher.path)
+	case err := <-doneCh:
+		return err
+	}
+}
+
+func (watcher *PathWatcher) consumeEvents(ctx context.Context, eventCh <-chan *gohan_sync.Event, watchErr chan<- error) {
+	watcher.updateCounter(1, "running")
+	defer watcher.updateCounter(-1, "running")
+
+	var err error
+	defer func() {
+		watchErr <- err
+	}()
+
+	for {
+		select {
+		case event, ok := <-eventCh:
+			if !ok {
+				err = ctx.Err()
+				return
+			}
+			if err = watcher.consumeEvent(event); err != nil {
+				return
+			}
+		case <-ctx.Done():
+			err = ctx.Err()
+			return
+		}
+	}
+}
+
+func (watcher *PathWatcher) consumeEvent(event *gohan_sync.Event) (err error) {
+	// each event & recovery must not be interrupted,
+	// otherwise we could easily end up in an inconsistent state
+	// if ctx is canceled either in running extensions or just before recovery
```

Comment:
On the other hand, if one of the functions called here never returns, we will never exit from `func (watcher *PathWatcher) run(` as it would wait forever on `doneCh`

Reply:
as discussed offline: there's no easy way out: in a hypothetical application build on Gohan framework, executing extensions may take an arbitrary amount of time.

let me extend the comment to mention this issue